File: sdks/js/src/opspec/interpreter/opcall/params/param/number/validate.ts
import Ajv from 'ajv'
const ajv = new Ajv()

// add formats not included in JSON schema standard
ajv.addFormat(
  'integer',
  {
    type: 'number',
    validate: (value:string) => /^[0-9]+$/.test(value)
  } as any
  )

/**
 * validates value against constraints
 * @param {Number} value
 * @param {Object} [constraints]
 * @return {Array<Error>}
 */
export default function validate(
  value: number,
  constraints?: any
) {
  constraints = Object.assign({ type: 'number' }, constraints)

  ajv.validate(constraints,
    value)
  return ajv.errors || []
}

File: sdks/go/internal/jsonschema/validate.go
package jsonschema

import (
	"github.com/santhosh-tekuri/jsonschema/v6"
)

// Validate validates a value against constraints
func Validate(
	value any,
	schema map[string]interface{},
) []error {
	// guard no schema
	if schema == nil {
		return nil
	}

	// register custom format checkers
	// gojsonschema.FormatCheckers.Add("docker-image-ref", formats.DockerImageRefFormatChecker{})
	// gojsonschema.FormatCheckers.Add("integer", formats.IntegerFormatChecker{})
	// gojsonschema.FormatCheckers.Add("semver", formats.SemVerFormatChecker{})
	schemaID := "schema.json" // Arbitrary ID for the schema
	compiler := jsonschema.NewCompiler()

	if err := compiler.AddResource(schemaID, schema); err != nil {
		return []error{err}
	}

	s, err := compiler.Compile(schemaID)
	if err != nil {
		return []error{err}
	}

	errorMessages := map[string]string{}
	extractErrorMessages(schema, "", errorMessages)

	// Validate the parsed document
	if err := s.Validate(value); err != nil {
		if vErr, ok := err.(*jsonschema.ValidationError); ok {
			var errs []error
			for _, cause := range vErr.Causes {
				errs = append(errs, cause)
			}
			return errs
		}
		return []error{err}
	}

	return nil
}
